### PR TITLE
Generalize DeepSpeech estimator and asr attack to support espresso

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
This PR generalizes codes to support espresso estimator.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/speech_recognition/pytorch_deep_speech.py [hunk 1]
```diff
@@ -676,6 +724,43 @@ def _transform_model_input(
...
+        """
+        self.model.train()
+
+    def batch_norm(self, train: bool) -> None:
+        """
+        Set all batch normalization layers into train or eval mode.
+
+        :param train: False for evaluation mode.
+        """
+        self.set_batchnorm(train=train)
```
Comment (beat-buesser) [inline art/estimators/speech_recognition/pytorch_deep_speech.py:762]: What is the reason for this method? It just seems to call the existing method.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @minhitbk Thank you very much for your pull request. I think it looks good. I have only one question. What do you think?

**Comment by YiwenShaoStephen** [comment]
Hi @minhitbk Thanks for this great update which makes imperceptible attack work more generally for other estimators.
I just have a confusion that do you think the windowing parameters used in this attack should be tied with those used in estimators? 
It seems to me that these windowing parameters can be fixed with the attack itself since it's targeting on Masking effects for spectrogram and could be a general function for audio samples without involving with the exact model. 
And also in practice, the feature extractors (e.g. fbank, stft, mfcc) may vary in different speech recognition models (some of them might don't have a one) so it would be very tricky if we want to support other models in the future.
Please correct me if I misunderstanding something here.

File: art/estimators/speech_recognition/pytorch_deep_speech.py (hunk 1 above)
Comment (minhitbk) [reply art/estimators/speech_recognition/pytorch_deep_speech.py:762]: @beat-buesser : We need to call this in the attack for the Deepspeech estimator, but the Espresso one does not need to call set_batchnorm, so they need to provide an empty method. Please note that the set_batchnorm is inherited from the PytorchEstimator where Espresso also inherits, so we cannot call this method directly.

**Comment by minhitbk** [comment]
@YiwenShaoStephen : Thanks very much for your comment. I agree with you that those window params should be part of the attack instead of the estimator. I discussed with @beat-buesser , so we will implement those params as attack's attributes which are passed to the attack through its `__init__` function. In the meanwhile, you can assume that you could get those from attack's self and don't need to wait us. We will update this PR soon. Thanks.

**Review by beat-buesser — APPROVED** [review]
Hi @minhitbk Thank you very much!
